Teach SubstitutionMap::isIdentity about non-canonical generic params

jrose-apple opened the pull request:
Do a weaker check here that only looks at the canonical generic params and guarantees that *those* substitute to themselves. There may be replacement types for other generic params too, to canonicalize them, but that's not a problem.

This fixes a crash trying to mangle decls with opaque result types that have generic signatures that canonicalize away a generic parameter.

rdar://problem/51775857

jrose-apple commented:
Can you think of any other interesting test cases that I didn't include?

@swift-ci Please test

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -684,13 +684,23 @@ void SubstitutionMap::profile(llvm::FoldingSetNodeID &id) const {
+      return;
+
+    unsigned ordinal = sig->getGenericParamOrdinal(paramTy);
+    auto replacement = getReplacementTypes()[ordinal]->getCanonicalType(sig);
```
I think it would be clearer if you just did Type(paramTy).subst(this)

jrose-apple replied:
That's not the same check, is it? That would also accept SubstitutionMaps that didn't have an entry for that parameter.

slavapestov replied:
SubstitutionMaps are always "complete". There's an entry for each parameter

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -684,13 +684,23 @@ void SubstitutionMap::profile(llvm::FoldingSetNodeID &id) const {
+  });
+
+  return !hasNonIdentityReplacement &&
+         countOfGenericParams == getReplacementTypes().size();
```
Is it ever possible for countOfGenericParams to not equal getReplacementTypes().size()? Maybe it should be an assertion instead?

jrose-apple replied:
I was wondering that, but I didn't know if it was safe. All valid SubstitutionMaps should behave like that, but what about a degenerate one?

slavapestov replied:
See above :)

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -684,13 +684,25 @@ void SubstitutionMap::profile(llvm::FoldingSetNodeID &id) const {
+      return;
+
+    auto replacementTy = Type(paramTy).subst(*this);
+    if (replacementTy->getCanonicalType() != paramTy->getCanonicalType())
```
I don't know why I wrote it this way, hang on.

jrose-apple committed:
Teach SubstitutionMap::isIdentity about non-canonical generic params

Do a weaker check here that only looks at the canonical generic params
and guarantees that *those* substitute to themselves. There may be
replacement types for other generic params too, to canonicalize them,
but that's not a problem.

This fixes a crash trying to mangle decls with opaque result types
that have generic signatures that canonicalize away a generic
parameter.

rdar://problem/51775857

jrose-apple commented:
@swift-ci Please smoke test
